mixin/smarty-v2@1/example/tests/mixin/SmartyTest.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Civi\Shimmy\Mixins;
+
+/**
+ * Assert that the 'scan-classes' mixin is working properly.
+ *
+ * This class defines the assertions to run when installing or uninstalling the extension.
+ * It is called as part of E2E_Shimmy_LifecycleTest.
+ *
+ * @see E2E_Shimmy_LifecycleTest
+ */
+class SmartyTest extends \PHPUnit\Framework\Assert {
+
+  public function testPreConditions($cv) {
+    $this->assertFileExists(static::getPath('/templates/CRM/Shimmy/Example.tpl'), 'The shimmy extension must have example TPL files.');
+  }
+
+  public function testInstalled($cv) {
+    $out = $this->renderExample($cv, 'CRM/Shimmy/Example.tpl');
+    $this->assertEquals('<p>OK</p>', trim($out));
+  }
+
+  public function testDisabled($cv) {
+    if ($cv->isLocal()) {
+      // Historically, Smarty templates have been left active for duration of same-process (post-disabling).
+      // We'll ignore testing this edge-case until someone decides that a change in behavior is better.
+      return;
+    }
+
+    $out = $this->renderExample($cv, 'CRM/Shimmy/Example.tpl');
+    $this->assertEquals('', trim($out));
+  }
+
+  public function testUninstalled($cv) {
+    // Same as disabled....
+    $this->testDisabled($cv);
+  }
+
+  protected static function getPath($suffix = ''): string {
+    return dirname(__DIR__, 2) . $suffix;
+  }
+
+  /**
+   * Render a template with the system-under-test.
+   *
+   * @param $cv
```

Comment:
I'm not sure what `$cv` refers to but it looks like it is an object from the code

Reply:
Yeah, this is a thing that's shared by all the mixin tests (`ls mixin/*/example/tests/mixin/`)

There are two implementations of this object -- [one sends requests to CiviCRM by calling cv CLI](https://github.com/civicrm/civicrm-core/blob/45a4c6ea449e969924e41c0eb9636d92e547aaad/tests/extensions/shimmy/tests/phpunit/E2E/Shimmy/LifecycleTest.php#L124), and [the other sends requests with local function calls](https://github.com/civicrm/civicrm-core/blob/45a4c6ea449e969924e41c0eb9636d92e547aaad/tests/extensions/shimmy/tests/phpunit/E2E/Shimmy/LifecycleTest.php#L100).

Why? The general idea is that we have different workflows in managing extensions -- sysadmin workflows tend to spawn separate PHP processes for every step (`cv en mymodule` or `POST /civicrm/admin/extension/...`), and unit-tests tend do all their steps in the same PHP process (`civicrm_api3('Extension', 'install', ...)`). It's a bit frustrating when a feature is written and works well with one workflow - but then fails in the other workflow (*because of some subtle difference caching*). The purpose of `testLifecycleWithSubprocesses()` and `testLifecycleWithLocalFunctions()` is to ensure that mixins are enabled/disabled consistently in both flows.

The name `$cv` might be a misnomer. It's more like "give me some channel to send commands to Civi, and I don't care how they get there". I've definitely been tempted to extract/formalize that. There are other channels (like HTTP and piped JSON-RPC). Hypothetically, the formalization might be:

```php
namespace Civi\Client;
interface ClientInterface {
  public function api3(string $entity, string $action, array $params): array;
  public function api4(string $entity, string $action, array $params): array;
  public function phpEval(string $code): mixed;
}
```

It shouldn't be a blocker for this PR (*since it's consistent with the style of other mixin tests; and since it is important for mixins to behave consistently in different workflows*). But if there's interest, I'm happy to pursue that. 

(*If formalizing, then there are some follow-ups/sub-topics... Does `ClientInterface` live in `civicrm-core.git` or a separate repo? Also, `phpEval()` is great for test-code - but how do we signal that it's channel-dependent and not recommended for live code?*)

Reply:
Yeah - tbh I did comment on the line here `* @param $cv` cos it was the comment block that was not declaring what type it was. I hadn't realised the `cv` was referring to the `cv` command line utility - so I don't know if the name is a misnomer - I'd mostly been at the level of 'is cv a string or an object or an integer'

Reply:
OK, fiar enough. Added the `object` hint for `$cv`.